lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -121,6 +123,24 @@ protected Object readArray(Object old, Schema expected, ResolvingDecoder in)
       @SuppressWarnings("unchecked")
       Collection<Object> c = (Collection<Object>) array;
       return readCollection(c, expectedType, l, in);
+    } else if (array instanceof Map) {
+      // Only for non-string keys, we can use NS_MAP_* fields
+      // So we check the samee explicitly here
+      if (ReflectData.isNonStringMapSchema(expected)) {
+        Collection<Object> c = new ArrayList<Object> ();
+        readCollection(c, expectedType, l, in);
+        Map m = (Map)array;
+        for (Object ele: c) {
+          IndexedRecord rec = ((IndexedRecord)ele);
+          Object key = rec.get(ReflectData.NS_MAP_KEY_INDEX);
+          Object value = rec.get(ReflectData.NS_MAP_VALUE_INDEX);
+          m.put (key, value);
+        }
+        return array;
+      } else {
+        throw new AvroRuntimeException("Type of " + array +
```

Comment:
Could you be more specific? There's no need to include the array here, since it is an empty map. And it should specifically state that it expected the element schema to be an array of non-string key and value pairs (or records).